### ISSUE TITLE
provide tooling similar to AFL++ for building PUTs

### ISSUE THREAD

**Issue by michaelmera** [tlspuffin/tlspuffin#308]
The idea would be to provide a drop-in compiler wrapper that performs the instrumentation of the PUT within its native build. This would be similar to [LibAFL CC](https://github.com/AFLplusplus/LibAFL/tree/main/libafl_cc) provided by AFL++.

**NOTE** Originates in [this PR review](https://github.com/tlspuffin/tlspuffin/pull/305#discussion_r1544581921):

> I think it would be great to have an even more manual process, where you only produce a static Rust library from tlspuffin and then call the linker on your own.
> 
> That way it would also possible to distribute binaries and add tooling around it that follow the classical approach of instrumenting/fuzzing.
> 
> In AFL++ for instance you have afl-cc that you can use to compile your program. After compilation it is no longer a program but a fuzzing target that can be executed. Much like clang -fsanitize=fuzzer. If we split all of the steps sufficiently, we can have a tlspuffin-cc that you can set using CC=tlspuffin-cc and then compile e.g. a patched OpenSSL.
>

**Comment by michaelmera** [comment]
A few remarks about this:

1. unlike AFL++ we don't target arbitrary binaries/libraries, but protocol implementations. This means that we need to completely orchestrate the execution, not just instrument the binary and link a fuzzer runtime. This implies knowledge about the instrumented libraries (APIs to create agents, ...).

2. because we don't target arbitrary binaries, it also means that the number of users that would use such tooling is limited and we already provide wrappers to several projects. It is not obvious to me that library developers that we support will want to spend the time to integrate tlspuffin upstream when they can just change the sources URL in the tlspuffin repository and let our infrastructure handle the heavy lifting.

3. we usually also want to patch the sources to insert the implementation for claims, querying internal knowledge about the PUT (i.e. mostly not accessible via provided APIs). For now this cannot be automated and worse: it is library AND library-version dependent. It means that we would need to integrate inside this compiler wrapper a bit of knowledge about ALL the PUTs that we support.

4. I don't see any benefit of this tooling for us developers of tlspuffin, because we need to setup the PUT harness anyway. So it is exclusively useful for external users, which at this point in time might be premature.

That being said, I concur that this is a terrific idea if we want to push the integration of tlspuffin upstream (in library repositories) as well as in large-scale fuzzing infrastructures. That would offload the maintenance of PUT-specific code to the relevant projects and allow us to focus on improving the fuzzer itself. However, it will take time and effort to convince upstream libraries maintainers and help setup the necessary changes.

Fundamentally, it seems to me that the choice of developing this tooling or not boils down to how we want tlspuffin to be perceived: is it a fuzzing runtime (a library, like AFL++) or is it a fuzzing infrastructure (a framework, like oss-fuzz)? Note that developing one does not exclude the other, but I would like that we discuss this issue with that in mind, because we also have limited resources to allocate to this development.

**Comment by LCBH** [comment]
> A few remarks about this:
> 
>     1. unlike AFL++ we don't target arbitrary binaries/libraries, but protocol implementations. This means that we need to completely orchestrate the execution, not just instrument the binary and link a fuzzer runtime. This implies knowledge about the instrumented libraries (APIs to create agents, ...).
> 
>     2. because we don't target arbitrary binaries, it also means that the number of users that would use such tooling is limited and we already provide wrappers to several projects. It is not obvious to me that library developers that we support will want to spend the time to integrate tlspuffin upstream when they can just change the sources URL in the tlspuffin repository and let our infrastructure handle the heavy lifting.
> 
>     3. we usually also want to patch the sources to insert the implementation for claims, querying internal knowledge about the PUT (i.e. mostly not accessible via provided APIs). For now this cannot be automated and worse: it is library AND library-version dependent. It means that we would need to integrate inside this compiler wrapper a bit of knowledge about ALL the PUTs that we support.
> 
>     4. I don't see any benefit of this tooling for us developers of tlspuffin, because we need to setup the PUT harness anyway. So it is exclusively useful for external users, which at this point in time might be premature.
> 
> 
> That being said, I concur that this is a terrific idea if we want to push the integration of tlspuffin upstream (in library repositories) as well as in large-scale fuzzing infrastructures. That would offload the maintenance of PUT-specific code to the relevant projects and allow us to focus on improving the fuzzer itself. However, it will take time and effort to convince upstream libraries maintainers and help setup the necessary changes.
> 
> Fundamentally, it seems to me that the choice of developing this tooling or not boils down to how we want tlspuffin to be perceived: is it a fuzzing runtime (a library, like AFL++) or is it a fuzzing infrastructure (a framework, like oss-fuzz but specific to protocol fuzzing)? Note that developing one does not exclude the other, but I would like that we discuss this issue with that in mind, because we also have limited resources to allocate to this development.

I agree that an `afl-cc` in the very general case seems unrealistic since part of the harnessing is still a manual process at the moment. However, in the short/medium-term, this idea could  become realistic for PUTs exposing an "OpenSSL compatibility layer", e.g., as WolfSSL offers. Indeed, the harness could solely use this compatibility layer and therefore be independent of the PUT.

**Comment by maxammann** [comment]
@michaelmera Thanks! This is on point:

> is it a fuzzing runtime (a library, like AFL++) or is it a fuzzing infrastructure (a framework, like oss-fuzz but specific to protocol fuzzing)? Note that developing one does not exclude the other, but I would like that we discuss this issue with that in mind, because we also have limited resources to allocate to this development.

I do agree that a very generic `tlspuffin-cc` might be overengineered an not a lot of users would use it. However, I actually did not want to push too much for that with my original discussion. Instead I believe, we should make sure that there is a clear way on how to fuzz a new implementation. You could translate my concern to: It would be great to have an easy to understand "baby fuzzer" with tlspuffin that can be understood by the fuzzing community.

The classical way would be to compile the following three components independently and then link them together manually:

* fuzzer engine (Rust)
* harness
* (optionally patched) PUT

I think it is very important to allow the user to use tlspuffin in that way.

I believe (and I certainly don't know the exact plans for the fuzzing project) this discussion came up from my end when I saw that Rust is still being used for linking and/or the general setup is still slightly complicated. Ideally you could compile tlspuffin once as static library and then reuse that across several PUTs and harnesses.

The addition of `tlspuffin-cc` is absolutely optional and a cherry on top.
